examples/ls-files.c

```diff
@@ -0,0 +1,143 @@
+ * This example demonstrates the libgit2 index APIs to roughly
+ * simulate the output of `git ls-files`.
+ * `git ls-files` has many options and this currently does not show them.
+ *
+ * `git ls-files` base command shows all paths in the index at that time.
+ * This includes staged and committed files, but unstaged files will not display.
+ *
+ * This currently supports the default behavior and the `--error-unmatch` option.
+ */
+
+typedef struct {
+	int error_unmatch;
+	char * files[1024];
+	int file_count;
+} ls_options;
+
+/* Print a usage message for the program. */
+static void usage(const char *message, const char *arg)
+{
+	if (message && arg)
+		fprintf(stderr, "%s: %s\n", message, arg);
+	else if (message)
+		fprintf(stderr, "%s\n", message);
+	fprintf(stderr, "usage: ls-files [--error-unmatch] [--] [<file>...]\n");
+	exit(1);
+}
+
+static int parse_options(ls_options *opts, int argc, char *argv[])
+{
+	int parsing_files = 0;
+	int i;
+
+	memset(opts, 0, sizeof(ls_options));
+
+	if (argc < 2)
+		return 0;
+
+	for (i = 1; i < argc; ++i) {
+		char *a = argv[i];
+
+		/* if it doesn't start with a '-' or is after the '--' then it is a file */
+		if (a[0] != '-' || parsing_files) {
+			parsing_files = 1;
+
+			// watch for overflows (just in case)
+			if (opts->file_count == 1024) {
+				break;
+			}
+
+			opts->files[opts->file_count++] = a;
+		} else if (!strcmp(a, "--")) {
+			parsing_files = 1;
+		} else if (!strcmp(a, "--error-unmatch")) {
+			opts->error_unmatch = 1;
+		} else {
+			usage("Unsupported argument", a);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int print_paths(ls_options *opts, git_index *index)
+{
+	int i;
+	const git_index_entry *entry;
+
+	/* loop through the files found in the args and print them if they exist */
+	for (i = 0; i < opts->file_count; ++i) {
+		const char *path = opts->files[i];
+
+		entry = git_index_get_bypath(index, path, GIT_INDEX_STAGE_NORMAL);
+		if (!entry && opts->error_unmatch) {
+			fprintf(stderr, "error: pathspec '%s' did not match any file(s) known to git.\n", path);
+			fprintf(stderr, "Did you forget to 'git add'?\n");
+			return -1;
+		}
+
+		printf("%s\n", path);
+	}
+
+	return 0;
+}
+
+int main(int argc, char *argv[])
+{
+	ls_options opts;
+	git_repository *repo = NULL;
+	git_index *index = NULL;
+	const git_index_entry *entry;
+	size_t entry_count;
+	size_t i = 0;
+	int error;
+
+	if ((error = parse_options(&opts, argc, argv)) < 0)
+		return error;
+
+	git_libgit2_init();
+
+	if ((error = git_repository_open_ext(&repo, ".", 0, NULL)) < 0)
+		goto cleanup;
+
+	if ((error = git_repository_index(&index, repo)) < 0)
+		goto cleanup;
+
+	/* if there are files explicitly listed by the user, we need to treat this command differently */
+	if (opts.file_count > 0) {
+		error = print_paths(&opts, index);
+	} else {
+		entry_count = git_index_entrycount(index);
+
+		for (i = 0; i < entry_count; i++) {
+			entry = git_index_get_byindex(index, i);
+			printf("%s\n", entry->path);
+		}
+	}
```

Comment:
Just a small style issue: I do find it a bit funny that some printing logic is in `print_paths` and some is not. Maybe put the complete priting logic into `print_paths` and destinguish cases by `opts.file_count` inside of that function, then? That would make the `main` function a bit nicer to read.